text/0000-edition-2021.md

```diff
@@ -0,0 +1,226 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Edition cadence not specified
+
+This RFC explicitly does not attempt to specify the cadence for releasing editions. Historically, they have occurred every 3 years, if one considers Rust 1.0 (released in 2015) to be the "original edition". Our expectation is that after Rust 2021 is shipped, we will use the experiences from Rust 2018 and Rust 2021 to review how editions are working and potentially establish a cadence for future editions (or perhaps make other changes to the edition system). 
+
+There are various constraints on the timing of editions that have been identified over time:
+
+* Even if it's automated, migrating to a new edition can require a significant investment of time, particularly for production users with many crates. We don't wish to release Editions at a rate that makes it hard for such users to keep up.
+* At the same time, we don't wish to wait too long in between editions. We want to maintain the spirit of Rust's train model, so that folks don't feel a "rush" to get work done by any particular deadline.
+* Having a generally known time when editions will be released is helpful in planning feature development.
+* We do not wish to release an empty edition. Sometimes there simply won't be any backwards incompatible changes and thus no reason to issue an edition.
+
+## Definition: migration
+
+**Migrations** are the "breaking changes" introduced by the edition, except of course that since editions are opt-in, no code actually breaks. 
+
+## Definition: migration lint
+
+Migrations typically come with one or more **migration lints**. Each lint warns about code that will need to change in order to move to the new edition. The lints typically contain "suggestions", which is a diff that can be applied to make the code work in the new edition. In some cases, the rewrite may be too complex for the lint to make a suggestion. In addition, some suggestions are marked as "not machine applicable", if they are not known to be semantics preserving.
+
+## Tooling workflow
+
+The expected workflow for upgrading to a new edition is as follows:
+
+* **Prepare to upgrade:** Run `cargo fix --edition`: This will automatically prepare your code to be upgraded to next edition by applying any suggestions from the migration lints.
+    * For example, if your code is in Rust 2018, this will prepare you to move to the 2021 edition.
+    * Note that `cargo fix` does not actually move your code to the new edition. Instead, it produces code that works in both the old and the new edition.
+        * This allows people to upgrade and fix things in a "piecemeal" fashion. Because of this, however, the code produces by these suggestions is not always the most idiomatic, as it is not able to take advantage of features from the new edition.
```

Comment:
```suggestion
        * This allows people to upgrade and fix things in a "piecemeal" fashion. Because of this, however, the code produced by these suggestions is not always the most idiomatic, as it is not able to take advantage of features from the new edition.
```